### PR TITLE
[23.2] Fix creating new erronous workflow can produce multiple empty workflows

### LINKED ISSUE #17405: [23.2] Creating new erronous workflow can produce multiple empty workflows
If you try to create a new workflow that has an error in it, you do not get an alert and can continuously click the save button, creating multiple empty workflows. **This could be related to https://github.com/galaxyproject/galaxy/pull/17330**

Galaxy Version: 23.2 (main)
Commit: [14a50465840f26283cc7c34165d9567632ec9aaf](https://github.com/galaxyproject/galaxy/commit/14a50465840f26283cc7c34165d9567632ec9aaf)

**To Reproduce**
1. Go to workflows list and click Create
2. Add steps that have an error (e.g: Duplicate labels)
3. Click the Save button (tooltip="Create a new workflow")
4. Though the workflow (without any steps) will be created, you will get no indication and no error modal.

**Expected behavior**
There must be an error message indicating what the error is, and the user mustn't be able to save the workflow at that point.

https://github.com/galaxyproject/galaxy/assets/78516064/ebdf71f8-da6f-4e8f-9e0c-08e10044d930

### PR DESCRIPTION
**Pull request by ahmedhamidawan**
Use `POST api/workflows` to create and save-as workflows, instead of the `workflows/create` and `workflows/save_workflow_as` controllers. This prevents the bug of creating (or saving as) workflows with errors and getting no prompts.
On the front-end, handle errors better and show them in editor error modal.

Fixes https://github.com/galaxyproject/galaxy/issues/17405
Fixes https://github.com/galaxyproject/galaxy/issues/17130

### Before:
https://github.com/galaxyproject/galaxy/assets/78516064/ebdf71f8-da6f-4e8f-9e0c-08e10044d930

### After:
https://github.com/galaxyproject/galaxy/assets/78516064/72b2feda-bb34-4cc7-a6af-9d8fd1257c1c

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by mvdbeek** [comment]
That's a regression in 23.2 that doesn't seem properly resolved here:
Screenshot 2024-02-01 at 10 57 22

This is not the only exception you can encounter, so it looks like the request handling on the client is maybe wrong ? Not saying we should ship html, but this doesn't look like a complete fix.

File: lib/galaxy/webapps/galaxy/controllers/workflow.py [hunk 1]
```diff
@@ -432,15 +428,29 @@ def save_workflow_as(
...
+                missing_tools_return = None
+
+            # Persist
+            session = trans.sa_session
+            session.add(stored_workflow)
+            with transaction(session):
+                session.commit()
+
+            if missing_tools_return:
+                return missing_tools_return
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/controllers/workflow.py]: Why did you make that change, it doesn't look related or correct.

**Comment by mvdbeek** [comment]
Right, so we previously used the API and now we use the controller, that's a step backwards that needs fixing.

File: lib/galaxy/webapps/galaxy/controllers/workflow.py (hunk 1 above)
Comment (ahmedhamidawan) [reply lib/galaxy/webapps/galaxy/controllers/workflow.py]: I moved the Persist after error checking 
for Duplicate labels because if the Persist is before the update function call, we create a new empty workflow, then add the steps, and then tell the user that their workflow has errors. This way, the new workflow is only created if we don't have the Duplicate label error.

However, for missing tools, we used to create the new workflow without the steps and then tell the user that their workflow has missing tools. So, I created the `missing_tools_return` variable to make sure we don't return without Persisting in the case of the missing tools error.

_I can make this treat the Duplicate label error the same way as we treat the Missing tools error if we're ok with that?_

Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/controllers/workflow.py]: I would prefer that we focus on the bug before refactoring things. I think that's not a clearcut improvement.

Comment (ahmedhamidawan) [reply lib/galaxy/webapps/galaxy/controllers/workflow.py]: But in order to fix this bug, we need to catch this exception

Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/controllers/workflow.py]: No, you need to use the API. This used to work fine on 23.1

Comment (ahmedhamidawan) [reply lib/galaxy/webapps/galaxy/controllers/workflow.py]: Ok, I'll look at what we did in 23.1, and try to use the api instead of the controller here.

File: lib/galaxy/webapps/galaxy/controllers/workflow.py [hunk 2]
```diff
@@ -432,15 +428,29 @@ def save_workflow_as(
                     workflow_data,
                     workflow_update_options,
                 )
+            except exceptions.DuplicatedIdentifierException as e:
+                return self.message_exception(trans, e.err_msg, False)
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/controllers/workflow.py]: This is only necessary because you're using the controller method, which you don't need to use. The API handles this correctly, `DuplicatedIdentifierException` derives from MessageException.

Comment (ahmedhamidawan) [reply lib/galaxy/webapps/galaxy/controllers/workflow.py]: Ah ok, now I understand why the exception wasn't being caught at all

File: lib/galaxy/webapps/galaxy/controllers/workflow.py (hunk 1 above)
Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/controllers/workflow.py]: What you want to do is replace `save_workflow_as` with a PUT to `/api/workflows/{id}` or a POST to `/api/workflows` (depending on whether the workflow has been saved before).

Comment (ahmedhamidawan) [reply lib/galaxy/webapps/galaxy/controllers/workflow.py]: Done, used `POST api/workflows` instead for both creating and saving-as workflows. It works and the errors are correctly handled at the client as well.

**Comment by ahmedhamidawan** [comment]
Failed tests:
```
FAILED test/integration_selenium/test_trs_import.py::TestTrsImport::test_import_workflow_by_url_workflowhub - selenium.common.exceptions.TimeoutException: Message: Timeout waiting on CSS selector [#workflow-table] to become visible.

FAILED test/integration_selenium/test_trs_import.py::TestTrsImport::test_import_by_id_workflowhub - AssertionError: workflow $COVID-19: variation analysis on ARTIC PE data not imported
assert 0 == 1

FAILED test/integration_selenium/test_trs_import.py::TestTrsImport::test_import_by_trs_url_workflowhub - AssertionError: workflow $COVID-19: variation analysis on ARTIC PE data not imported
assert 0 == 1
```

seem unrelated to the changes in this PR.

File: client/src/components/Workflow/Editor/Index.vue [hunk 3]
```diff
@@ -532,30 +536,27 @@ export default {
...
-            const formData = new FormData();
-            formData.append("workflow_name", rename_name);
-            formData.append("workflow_annotation", rename_annotation);
-            formData.append("from_tool_form", true);
-            formData.append("workflow_data", JSON.stringify(toSimple(this.id, this)));
+        async doSaveAs() {
+            if (!this.saveAsName && !this.nameValidate()) {
+                return;
+            }
+            const rename_name = this.saveAsName ?? `SavedAs_${this.name}`;
```
Comment (mvdbeek) [inline client/src/components/Workflow/Editor/Index.vue:543]: The `SavedAs` case can't actually happen because of the check on line 540, or am I misreading this ?

**Review by mvdbeek — APPROVED** [review]
Looks awesome, thanks so much!

File: client/src/components/Workflow/Editor/Index.vue (hunk 3 above)
Comment (ahmedhamidawan) [reply client/src/components/Workflow/Editor/Index.vue:543]: You are right, i messed this up refactoring the Save as function, thanks for pointing this out

Comment (ahmedhamidawan) [reply client/src/components/Workflow/Editor/Index.vue:543]: Actually, no, that would be the case if line 540 had an OR

**line 540:**
- if user doesn't provide a _Save As name_ in the save as modal, 
- AND the user doesn't have a `this.name` (set in the Editor right panel name input field)

_return_

**Now in line 543:**
- it is still possible that the user didn't provide a _Save As name_, but they might have still had a `this.name`, so save the workflow as "SaveAs_OGWf name"
